### ISSUE TITLE
AI of Deepmind discovered 4x4 x 4x4 non-comm. multip. algorithm with just 47 mul.

### ISSUE THREAD

**Issue by ValZapod** [libxsmm/libxsmm#698]
https://www.nature.com/articles/s41586-022-05172-4

This of course will destroy all sceptics who did not believe that neural networks can do math in symbolic or even pure form. Considering oneMKL is only so fast because of this library (and by itself outperfoms everything else) I will report it here first and cause your library is called small matrix multiplication, 4x4 x 4x4 is indeed very small.

THIS ALGORITHM is **non-commutative**, i.e. has A elements on one side and B matrix elements on the other of EACH "*". THAT means it can be used recursively to multiply block matrices.

![41586_2022_5172_Fig6_ESM](https://user-images.githubusercontent.com/31514790/194192327-bb3eef24-5464-430c-98d8-47bfdfcbf529.jpg)

P.S. Kudos for ARM9 SVE2 patches. :) Still did not get how to run SVE2 on my Snap 8 gen 1 though.

**Comment by alheinecke** [comment]
The paper says:
`We train AlphaTensor to search for efficient algorithms to multiply
4 × 4 block matrices, and focus on square matrix multiplication of size
8,192 (each block is hence of size 2,048) to define the benchmarking
reward. AlphaTensor searches for the optimal way of combining the
16 square blocks of the input matrices on the considered hardware. We
do not apply the 4 × 4 algorithm recursively, to leverage the efficient
implementation of matrix multiplication on moderate-size matrices
(2,048 × 2,048 in this case).`
So the they are doing classic matmul of 2Kx2K and this not libxsmm.

**Comment by ValZapod** [comment]
2K is now small with the new algorithm. Just implement it.

**Comment by alheinecke** [comment]
this oneMKL field.

**Comment by ValZapod** [comment]
So 4x4 is oneMKL?

**Comment by alheinecke** [comment]
there is no 4x4 matmul in this paper. 4 x 4 block matmul and each block is 2K x 2K, so you have classic 2K x 2K matmul and matadds... these are both library calls to oneMKL. And instead of doing 64 2K x 2K matmuls, they are 47 2k x 2K matmuls and buch of 2k X 2xK matadds.

**Comment by ValZapod** [comment]
>hence the gains are not 47/64 = 27%, but 5-10ish.

No, they are not. You are not counting electricity. Multiplication is still skower than summation and fasterbthan divisition (div**q** 128 bit devision is faster on Ice Lake, nowadays, LOL, and does not require libdivide). It is no slower by IPC or latencies. It is slower because it requires more energy on Minix OS. You know, operaying system of your PCH on the motherboard.

**Comment by ValZapod** [comment]
Besides no one uses 64 mul. operations for 4x4 x 4x4. Compilers apply Strassen automatically, so only 49.

**Comment by ValZapod** [comment]
BTW, on math.stackechange people are saying this is only valid over mod2, i.e. over Z/2Z... 

https://math.stackexchange.com/questions/578342/number-of-elementary-multiplications-for-multiplying-4x4-matrices#4546116

**Comment by hfp** [comment]
> Besides no one uses 64 mul. operations for 4x4 x 4x4. Compilers apply Strassen automatically, so only 49.

I doubt any compiler does that. Though, matrix multiply algorithms with less than O(N^3) often hinder vectorization.

**Comment by hfp** [comment]
> > hence the gains are not 47/64 = 27%, but 5-10ish.
> 
> No, they are not. You are not counting electricity. Multiplication is still slower than summation and faster than division (div**q** 128 bit division is faster on Ice Lake, nowadays, LOL, and does not require libdivide, still worse than Apple M1). It is not slower by IPC or latencies, though. It is slower because it requires more energy on Minix OS. You know, operating system of your PCH on the motherboard.

Thank you reporting and sharing the paper!  Let's stick to facts not dissolving in an endless chain of arguments. Also, faster/slower is usually referred as time to solution. Metrics on top of "time to solution" (like "energy to solution") are of course of interest as well.

**Comment by ValZapod** [comment]
BTW, here is cool commutative code for 21 mul. for 3x3 x 3x3. From new paper: https://arxiv.org/abs/1904.07683

Looks like it is just a little slower than code gcc produces for naive 27 mul. algorithm with -O3 -fun-unroll, but I suppose that assembly is very hacky, as it is a pattern recognized by gcc.

https://gcc.godbolt.org/z/9oszvM9aG

**Comment by ValZapod** [comment]
All results. Both modular and normal arithmetic. 7 MB file. 
[equations.txt](https://github.com/libxsmm/libxsmm/files/9737591/equations.txt)

**Comment by ValZapod** [comment]
Here is code written for small matrices too https://github.com/paolodalberto/FastMMW/issues/1
